From fb583d49b367294e859d2b2f298108a35cafd7e0 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 20 Oct 2023 16:52:39 +0200
Subject: [PATCH 0417/1695] core: arm: get DDR range from embedded DTB

Find main memory (DDR) physical range(s) from the secure embedded DTB
if not found from the external DDR.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 84275c32afda0f0fe7d216ce60115640e619fa1c core: arm: get DDR range from embedded DTB]
Change-Id: Ie6104c3eedfd23ea22ff0e32d5e8a1219c0787a0
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/arch/arm/kernel/boot.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/kernel/boot.c b/core/arch/arm/kernel/boot.c
index 7dc1b3a76..29a4bb7a2 100644
--- a/core/arch/arm/kernel/boot.c
+++ b/core/arch/arm/kernel/boot.c
@@ -1039,7 +1039,18 @@ static void discover_nsec_memory(void)
 			return;
 		}
 
-		DMSG("No non-secure memory found in FDT");
+		DMSG("No non-secure memory found in external DT");
+	}
+
+	fdt = get_embedded_dt();
+	if (fdt) {
+		mem = get_nsec_memory(fdt, &nelems);
+		if (mem) {
+			core_mmu_set_discovered_nsec_ddr(mem, nelems);
+			return;
+		}
+
+		DMSG("No non-secure memory found in embedded DT");
 	}
 
 	mem_begin = phys_ddr_overall_begin;
-- 
2.47.2

